models/auth/oauth2.go

```diff
@@ -54,6 +55,9 @@ func (app *OAuth2Application) PrimaryRedirectURI() string {
 
 // ContainsRedirectURI checks if redirectURI is allowed for app
 func (app *OAuth2Application) ContainsRedirectURI(redirectURI string) bool {
+	if setting.OAuth2.EnableRedirectURIWildcard {
+		return util.StringMatchesAnyPattern(redirectURI, app.RedirectURIs, true)
```

Comment:
You may like `github.com/gobwas/glob` which is already used by Gitea.

Reply:
Ah, I'll take a look thanks

Reply:
Hm. I think the asterisk-only approach is preferable here since there's no way to disable some of the patterns in the `glob` package. For example, `?` is valid in URLs, but in the pattern it would glob to any single character.